README.md
inclinometer_solar2
=============
Package to configure and read the dual axis inclinometer SOLAR-2-45-RS232 with Temperature compensation from leveldevelopements.

[LINK](https://www.leveldevelopments.com/products/inclinometers/inclinometer-sensors/solar-2-45-2-rs232-dual-axis-inclinometer-45-rs232-interface-with-temperature-compensation/)

Configuration
-------------
It is possible to configure the sensor using the configuration software on Windows found [here](https://www.leveldevelopments.com/wp/wp-content/uploads/software/Inclinometer_App.zip).

The sensor can also be configured using any serial terminal like Teraterm, 
[cutecom](http://cutecom.sourceforge.net/), etc.


Setup
-----
Test the sensor by building the sensor_tester:

```amake```

and then running it:

```./build/src/sensor_tester```

The orogen component can be tested by building it:

```amake ../orogen/inclinometer_solar2```

and then running the following ruby script>

```./build/src/sensor_tester```

<!-- 
License
-------
dummy-license

Installation
------------
The easiest way to build and install this package is to use Rock's build system.
See [this page](http://rock-robotics.org/stable/documentation/installation.html)
on how to install Rock.

However, if you feel that it's too heavy for your needs, Rock aims at having
most of its "library" packages (such as this one) to follow best practices. See
[this page](http://rock-robotics.org/stable/documentation/packages/outside_of_rock.html)
for installation instructions outside of Rock.

Rock CMake Macros
-----------------

This package uses a set of CMake helper shipped as the Rock CMake macros.
Documentations is available on [this page](http://rock-robotics.org/stable/documentation/packages/cmake_macros.html).

Rock Standard Layout
--------------------

This directory structure follows some simple rules, to allow for generic build
processes and simplify reuse of this project. Following these rules ensures that
the Rock CMake macros automatically handle the project's build process and
install setup properly.

### Folder Structure

| directory         |       purpose                                                        |
| ----------------- | ------------------------------------------------------               |
| src/              | Contains all header (*.h/*.hpp) and source files                     |
| build/ *          | The target directory for the build process, temporary content        |
| bindings/         | Language bindings for this package, e.g. put into subfolders such as |
| ruby/             | Ruby language bindings                                               |
| viz/              | Source files for a vizkit plugin / widget related to this library    |
| resources/        | General resources such as images that are needed by the program      |
| configuration/    | Configuration files for running the program                          |
| external/         | When including software that needs a non standard installation process, or one that can be easily embedded include the external software directly here |
| doc/              | should contain the existing doxygen file: doxygen.conf               |
 -->

CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for inclinometer_solar2 package
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

TODO
------------------
* implement full check to see what message arrives in buffer.
* implement logger
* implement different responses than OK in sendCmd()
* add purge method
* change output to RADs
* change output to IMU message

KNOWN BUGS
------------------
* When the buffer is purged it is directly full again for some reason. Occurs at rate = 50 ms.

NICE TO HAVE
------------------
* Online configuration and zeroing


0.1 (2019-01-08)
------------------
* rate was removed from initialization as it can anyway only be changed after the serial port was started
* added methods to adjust parameters of sensor
* turns off continuous transmission to change settings
* parses sensor values to port
* test script start.rb works
* filled update hook and configuration hook
* started programming orogen component
* sending a command works and the respons can be read.
* managed to concatenate message from two uint8_t variables
* sending a test command to configure the sensor works
* fixed timing issue and buffer parsing
* parsing of serial readout implemented
* readout of continuous serial communication works

test/test_Dummy.cpp
#include <boost/test/unit_test.hpp>
#include <inclinometer_solar2/Dummy.hpp>

using namespace inclinometer_solar2;

BOOST_AUTO_TEST_CASE(it_should_not_crash_when_welcome_is_called)
{
    inclinometer_solar2::DummyClass dummy;
    dummy.welcome();
}


src/main.cpp
#include <iostream>
#include <cstring>
#include "Solar2.hpp"
#include <chrono>

float ui8tof(uint8_t *input) {
    std::string s;
    s.assign(input, input + sizeof(input));
    return std::stof(s);
}

int main(int argc, char** argv)
{

    printf("Welcome to the inclinometer test software\n");

    inclinometer_solar2::Solar2 inclinometer(50);

    std::string const port = "/dev/ttyUSB0";

    // TODO This error catcher doesn't seem to work...
    if(!inclinometer.openSerial(port, 38400))
    {
        std::cout << "Could not open device on port " << port << std::endl;
        return 0;
    }
    else printf("Port %s opened.\n", port.c_str());

    int action;

    printf("-----------------------------------------\n");
    printf("1. Display data to terminal.\n");
    printf("2. Change transmission rate.\n");
    scanf("%i", &action);

    float inclinations[2];

    switch (action) {

        case 1:
            while(true)
            {
            	inclinometer.update(inclinations);
            }
        
        case 2:
            int rate;

            // printf("Input Transmission Rate [50, 9999 ms]");
            // scanf("%i", &rate);
            inclinometer.setRate(rate);
            while(true)
            {
                inclinometer.update(inclinations);      
            }
    }


    printf("Program finished. \n");

    return 0;
}


src/Solar2.hpp
#ifndef _SOLAR2_HPP_
#define _SOLAR2_HPP_

#include <iodrivers_base/Driver.hpp>

// Use cutecom for Ubuntu Console

namespace inclinometer_solar2
{
    class Solar2: public iodrivers_base::Driver
    {
    public:
        Solar2();

        ~Solar2();

        bool update(float * inclinations);
        void activatePrinting() {print_stream = false;}
        bool setRate(int rate);
        // Sends a command using sendMsg() but deactivates the streaming first and activates the streaming afterwards again.
        // set expect_ok to true if the response should be an ok and otherwise provide the location of the response.
        bool sendCmd(std::string command,  bool expect_ok=false, uint8_t *response = NULL);
        // Sends the message and reads a response
        // TODO: Implement responses other than OK
        bool sendMsg(uint8_t *message,  bool expect_ok=false, uint8_t *response = NULL); // Sends a single message and checks if the answer is OK. Change expect_ok to false if the response is different than OK.

        // Functions for fd-driven tasks
        int getFileDescriptor();
        // int getIndex();

    private:
        static const unsigned int CASC_MSG_SIZE = 18;  // Number of bytes per message in streaming mode
        static const unsigned int MAX_PACKET_SIZE = 512; // Has to be >512
        mutable int removed_bytes = 0; // Number bytes that were removed during
        mutable bool purge_buffer = false;
        int rate_;

        // Plot Options
        bool print_stream = false;
        bool print_buffer = false;

        // Comm functions
        virtual int extractPacket(uint8_t const* buffer, size_t buffer_size) const;
        bool writeMessage();

        // Helper functions
        float ui8tof(uint8_t *input);   // parses a uint8_t ascii array into a float.
        uint8_t* strtoui8t(std::string input);
        int nrOfDigits(int input);

    };
}

#endif


src/sensor_tester.cpp
#include <iostream>
#include <cstring>
#include "Solar2.hpp"
#include <chrono>

float ui8tof(uint8_t *input) {
    std::string s;
    s.assign(input, input + sizeof(input));
    return std::stof(s);
}

int main(int argc, char** argv)
{

    printf("Welcome to the inclinometer test software\n");

    inclinometer_solar2::Solar2 inclinometer;

    std::string const port = "/dev/ttyS1";

    // TODO This error catcher doesn't seem to work...
    if(!inclinometer.openSerial(port, 38400))
    {
        std::cout << "Could not open device on port " << port << std::endl;
        return 0;
    }
    else printf("Port %s opened.\n", port.c_str());

    int action;
    float inclinations[2];

    while(true){
        printf("-----------------------------------------\n");
        printf("1. Display data to terminal.\n");
        printf("2. Change transmission rate.\n");
        printf("3. Send string command (works only for 'OK' responses atm).\n");
        printf("-----------------------------------------\n");
        
        scanf("%i", &action);


        switch (action) {

            case 1:
                inclinometer.activatePrinting();
                while(true)
                {
                    inclinometer.update(inclinations);
                }
                break;  

            case 2:
                int rate;

                printf("Input Transmission Rate [50, 9999 ms]:\n");
                scanf("%i", &rate);
                inclinometer.setRate(rate);
                break;  

            case 3:
                char command[256];

                printf("Input the command to be sent:\n");
                scanf( "%255s" , command);

                inclinometer.sendCmd(std::string(command), true);
                break;  
        }


    }

    printf("Program finished. \n");

    return 0;
}


src/Solar2.cpp
#include <iostream>
#include <chrono>
#include <string.h>

#include "Solar2.hpp"

using namespace inclinometer_solar2;

Solar2::Solar2():
	Driver(MAX_PACKET_SIZE)
{
}


Solar2::~Solar2()
{
    // Close the serial connection if it is still valid
    if(isValid())
    {
        close();
    }
}

int Solar2::getFileDescriptor()
{
    return Driver::getFileDescriptor();
}

bool Solar2::update(float * inclinations)
{
    uint8_t packet[MAX_PACKET_SIZE];

	auto start = std::chrono::high_resolution_clock::now();

	try
	{
		// Timeout at 10s as this is is lowest transmission rate in streaming mode
		readPacket(packet, MAX_PACKET_SIZE, 10000, 10000);

		if (print_stream)
		{

			auto stop = std::chrono::high_resolution_clock::now();
			auto duration = std::chrono::duration_cast<std::chrono::microseconds>((stop - start)/1000);

			// PRINT TIME
			// To get the value of duration use the count()
			// member function on the duration object
			printf("--------------\n");
			printf("Elapsed: %ld ms\n", duration.count());


			// PRINT MESSAGE
			printf("Received Message [°]: ");

			// find \r and replace with readable \r
			for (int j = 0; j < 18; j++)
			{
				if (packet[j] == '\r')	std::cout << "\\r";
				else std::cout << packet[j];
			}

			printf("\n");
		}

		// TODO: implement full check to see what message it is.
		// Parse the message into a float array if the message does not contain an OK.
		if(packet[0] != 'O' && packet[1] != 'K'){

			// Extract the x & y values from the message packet.
		 	uint8_t x[9];
	    uint8_t y[9];

	    std::copy(packet, packet + 8, x);
	    std::copy(packet + 9, packet + 18, y);

	    // Parse the uint8_t arrays into a a float array containing x & y.
			float incs[2] = {ui8tof(x), ui8tof(y)};

			// Copy data to provided address
			memcpy(inclinations, &incs, sizeof(incs));
		}

		return true;
	}

	catch(int timeout_error)
	{
        std::cout << "inclinometerSolar2_Driver: Timeout error" << std::endl;
        return false;
	}

}


bool Solar2::sendMsg(uint8_t *message, bool expect_ok, uint8_t *response)
{
	int count = 0;
    uint8_t packet[MAX_PACKET_SIZE];

	writePacket(message, sizeof(message)/sizeof(message[0])-1);
	if(expect_ok)
	{
		while (count < 100) {
			readPacket(packet, MAX_PACKET_SIZE, 2*rate_, 2*rate_);
			if(packet[0] == 'O' && packet[1] == 'K') return true;

			count++;
		}
		printf("ERROR: No response received!\n");
	}
	else
	{
		return true;
	}

	return false;
}

bool Solar2::sendCmd(std::string command, bool expect_ok, uint8_t *response)
{
	if(!sendMsg(strtoui8t("stpcasc"), true))
	{
		printf("ERROR: Couldn't stop the stream.\n");
		return false;
	}

	if(!sendMsg(strtoui8t(command), 	expect_ok))
	{
		printf("ERROR: Couldn't send the command.\n");
		return false;
	}

	if(!sendMsg(strtoui8t("setcasc"),false))
	{
		printf("ERROR: Couldn't start the stream.\n");
		return false;
	}

	return true;
}


// Early testing function for commands.
bool Solar2::setRate(int rate)
{
	if (rate >= 50 && rate <= 9999)
	{

		// Build message
		rate_ = rate;
		std::string rate_message = "str";

		for (int i=0; i<(4-nrOfDigits(rate)); i++) rate_message.append("0");

		rate_message.append(std::to_string(rate));


		if(sendCmd(rate_message,true))
		{
			printf("Rate successfully set to %i ms.\n", rate);
			return true;
		}

	}
	else
	{
		printf("ERROR: chosen rate out of bounds.\n");
	}

	return false;
}


// Virtual method, must be redefined to process custom packet
int Solar2::extractPacket(uint8_t const* buffer, size_t buffer_size) const
{

	if(print_buffer)
	{
		printf("BS:\t%lu \n", buffer_size);
		printf("BF:\t\t");
		for (uint j = 0; j < buffer_size; j++)
		{
			if (buffer[j] == '\r')	std::cout << "\\r";
			else std::cout << buffer[j];
		}
		printf("\n");
	}

	if(purge_buffer){
		purge_buffer = false;

		printf("\t\t\t\t\t\t\t\t%lu Elements removed from buffer.\n", buffer_size);
		return -buffer_size;
	}


	if(buffer[0] == 'O' && buffer[1] == 'K' && buffer_size >= 2)
	{
		// printf("\t\t\t\t\t\t\t\tReceived OK from sensor.\n");
		return 2;
	}

	// One packet must be at least as long as a message
	if(buffer_size < CASC_MSG_SIZE)
	{
		return 0;
	}

	// Each message paket is terminated with a carriage return (\r).
	// Once this is in the buffer at the last position (17) we know the whole message is in the buffer
	// and the message size is returned.
	if (buffer[17] == '\r')
	{
		if (print_stream)
		{

			// PRINT LOST BYTES
			printf("Discarded Bytes: %d\n", removed_bytes);
			removed_bytes = 0;
		}

		return CASC_MSG_SIZE;
	}

	removed_bytes = removed_bytes + 1;

	// Remove one byte at beginning of the device buffer
	return -1;
}


float Solar2::ui8tof(uint8_t *input) {
    std::string s;
    s.assign(input, input + sizeof(input));
    return std::stof(s);
}

uint8_t* Solar2::strtoui8t(std::string input) {

    uint8_t* output;
    output = new uint8_t[input.length() + 1];
    memcpy(output, input.c_str(), input.length() + 1);
    return output;
}

int Solar2::nrOfDigits(int input){

	int length = 1;
	while ( input /= 10 )
	   length++;
	return length;
}
